### PR TITLE
fix: Fixed invalid jar fingerprints for maven projects

### PR DESCRIPTION
**Pull request by pecodez**
- [ ] Ready for review
- [ ] Follows CONTRIBUTING rules
- [ ] Reviewed by Snyk internal team

#### What does this PR do?
Refactored the recursion of unpacking nested JARs to avoid mutating objects
which was difficult to follow and caused incorrect dependency allocation
in the fingerprints facts which resulted in an invalid dep-graph.

### CODE REVIEW

File: lib/analyzer/applications/java.ts [hunk 1]
```diff
@@ -172,39 +169,39 @@ function unpackJarsTraverse({
...
 }
 
+/**
+ * Manages the unpacking an array of JarBuffer objects and returns the resulting
+ * fingerprints. Recursion to required depth is handled here when the returned
+ * info from each JAR that is unpacked has nestedJars.
+ *
+ * @param {JarBuffer[]} jarBuffers
+ * @param {number} desiredLevelsOfUnpacking
+ * @param {number} unpackedLevels
```
Comment (noaleibo1) [inline lib/analyzer/applications/java.ts:198]: Curious about what this annotation means. Where is this loaded from? Some configuration file?

**Review by noaleibo1 — COMMENTED** [review]
Understood very little of this PR obviously, and even the part I did understand I am not sure about  😅 
Are we basically limiting the number of levels we dive into, to avoid the "overflow" error? If so, to what number of levels? And where is it configured?

File: lib/analyzer/applications/java.ts (hunk 1 above)
Comment (pecodez) [reply lib/analyzer/applications/java.ts:198]: Yes, we do apply a limit to the level of unpacking. By default, this is 1 level but controlled with the `--nested-jar-depth` flag to allow the customer control. We actually always unpack 1 level deeper than the `desiredLevelOfUnpacking` as that is how we find shaded JARs and coords for custom JARs. We only generate fingerprints for JARs found up to the `desiredLevelOfUnpacking` though.

File: lib/analyzer/applications/java.ts [hunk 2]
```diff
@@ -218,29 +215,41 @@ function unpackJars(
...
+      dependencies: jarInfo.dependencies,
+      ...jarInfo.coords,
     });
+
+    if (jarInfo.nestedJars.length > 0) {
+      // this is an uber/fat JAR so we need to unpack the nested JARs to
+      // analyse them for coords and further nested JARs (depth flag allowing)
+      fingerprints.push(
+        ...unpackJars(
+          jarInfo.nestedJars as JarBuffer[],
```
Comment (shaninja) [inline lib/analyzer/applications/java.ts:247]: does it really have all the properties of JarBuffer, or is it a partial? would it make more sense to change line 202 to `Partial<JarBuffers>[]` (if that's a thing...)? not sure I fully follow and understand the source of the type difference...

File: test/system/application-scans/java.spec.ts [hunk 3]
```diff
@@ -56,16 +56,22 @@ describe("jar binaries scanning", () => {
               pluginResult.scanResults[1].facts[0].data.fingerprints;
           });
 
-          it("should return two results", () => {
-            expect(fingerprints).toHaveLength(2);
+          it("should return three results", () => {
```
[the remark this answers is not in the document]
Comment (pecodez) [reply test/system/application-scans/java.spec.ts:59]: I agree, it isn't clear. I will update the test name to be clearer

File: lib/analyzer/applications/java.ts (hunk 2 above)
Comment (pecodez) [reply lib/analyzer/applications/java.ts:247]: This was super unclear too. I've reduced the scope of the `JarBuffer` type and extended the `JarInfo` type. Now `unpackJars` method takes a `JarBuffer[]` and `unpackJar` method returns `JarInfo` which has the decorated information needed to create a `Fingerprint` type.

File: test/system/application-scans/java.spec.ts [hunk 4]
```diff
@@ -251,7 +257,7 @@ describe("jar binaries scanning", () => {
             fingerprints =
               pluginResult.scanResults[1].facts[0].data.fingerprints;
 
-            expect(fingerprints).toHaveLength(1);
+            expect(fingerprints).toHaveLength(3);
```
[the remark this answers is not in the document]
Comment (pecodez) [reply test/system/application-scans/java.spec.ts:260]: The assertions have not actually changed; they both previously expected 1 fingerprint matching the deepest level. I believe this test case was more about checking things not breaking when specifying a depth greater than the actual levels of nested JARs and so have renamed the test to be specific.

[the remark this answers is not in the document]
Comment (shaninja) [reply test/system/application-scans/java.spec.ts:260]: ah, true. I think I got confused when reading the previous assertion.

**Review by shaninja — APPROVED** [review]

**Review by noaleibo1 — APPROVED** [review]

**Pull request merged**